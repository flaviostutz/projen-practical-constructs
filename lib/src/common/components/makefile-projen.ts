/* eslint-disable no-new */
import { Component, Project, SampleFile } from 'projen';

import { NvmRcFile } from '../../nodejs/files/nvmrc';

export class MakefileProjen extends Component {
  constructor(project: Project, opts?: MakefileProjenOptions) {
    super(project);

    const optsWithDefaults: Required<MakefileProjenOptions> = {
      tsNodeLibVersion: '10.9.2',
      projenLibVersion: '0.91.6',
      nodeVersion: '20.16.0',
      additionalContents: '',
      additionalContentsTargets: {},
      ...opts,
    };

    new SampleFile(project, 'Makefile', {
      contents: `all: build lint test

build:
      npx projen build
      ${optsWithDefaults.additionalContentsTargets.build ?? ''}

lint:
      npx projen lint
      ${optsWithDefaults.additionalContentsTargets.lint ?? ''}

lint-fix:
      npx projen lint-fix
      ${optsWithDefaults.additionalContentsTargets['lint-fix'] ?? ''}

test:
      npx projen test
      ${optsWithDefaults.additionalContentsTargets.test ?? ''}

clean:
      npx projen clean
      ${optsWithDefaults.additionalContentsTargets.clean ?? ''}

prepare:
      brew install nvm
      @echo "Configure your shell following the instructions at https://formulae.brew.sh/formula/nvm"
      ${optsWithDefaults.additionalContentsTargets.prepare ?? ''}

prepare-projen:
      npm install --no-save ts-node@${optsWithDefaults.tsNodeLibVersion} projen@${optsWithDefaults.projenLibVersion}

${optsWithDefaults.additionalContents}
`,
    });

    new NvmRcFile(project, {
      nodeVersion: optsWithDefaults.nodeVersion,
    });
  }
}

export interface MakefileProjenOptions {
  /**
   * Additional contents to be added to the Makefile on top of the default projen related targets
   * @default - no additional rules
   */
  readonly additionalContents?: string;
  /**
   * The version of ts-node lib to be used in "prepare" target of the Makefile
   * to install tooling for Projen to work
   * @default '10.9.2'
   */
  readonly tsNodeLibVersion?: string;
  /**
   * The version of projen lib to be used in "prepare" target of the Makefile
   * to install tooling for Projen to work
   * @default '0.91.6'
   */
  readonly projenLibVersion?: string;
  /**
   * Node version to be added to the .nvmrc file
   * @default '20.16.0'
   */
  readonly nodeVersion?: string;
  /**
   * Additional contents to be added to each target of the Makefile
   */
  readonly additionalContentsTargets?: Record<string, string>;
}
