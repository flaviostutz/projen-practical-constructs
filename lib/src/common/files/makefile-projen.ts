/* eslint-disable no-restricted-syntax */
import { FileBase, IResolver, Project } from 'projen';

import { TS_NODE_VERSION } from '../../python/constants';
import { NODE_VERSION, PROJEN_VERSION } from '../constants';

export class MakefileProjenFile extends FileBase {
  optsWithDefaults: Required<MakefileProjenOptions>;

  constructor(project: Project, opts?: MakefileProjenOptions) {
    super(project, 'Makefile-projen');

    this.optsWithDefaults = {
      tsNodeLibVersion: TS_NODE_VERSION,
      projenLibVersion: PROJEN_VERSION,
      nodeVersion: NODE_VERSION,
      additionalMakefileContentsProjen: '',
      additionalMakefileContentsTargets: {},
      ...opts,
    };
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  protected synthesizeContent(_resolver: IResolver): string | undefined {
    // eslint-disable-next-line functional/no-let
    let contents = `SHELL := /bin/bash

all: build lint test

prepare:
	brew install nvm
	@echo "Configure your shell following the instructions at https://formulae.brew.sh/formula/nvm"
  ${targetContents(this.optsWithDefaults.additionalMakefileContentsTargets.prepare)}

prepare-projen:
	@if [ "$$CI" == "true" ]; then \
		set -x; npm install --no-save --no-package-lock ts-node@${this.optsWithDefaults.tsNodeLibVersion} projen@${this.optsWithDefaults.projenLibVersion}; \
	else \
		set -x; npm install --no-save ts-node@${this.optsWithDefaults.tsNodeLibVersion} projen@${this.optsWithDefaults.projenLibVersion}; \
	fi

`;

    for (const task of this.project.tasks.all) {
      if (task.name === 'default') {
        // eslint-disable-next-line no-continue
        continue;
      }
      const additional =
        targetContents(this.optsWithDefaults.additionalMakefileContentsTargets[task.name]) ?? '';
      contents += `# ${task.description}
${task.name.replace(':', '-')}:
	npx projen ${task.name}${additional}

`;
    }
    return contents;
  }
}

const targetContents = (contents?: string): string => {
  if (!contents) return '';
  // add spaces to each line
  const contentWithSpaces = contents?.replace(/\n/g, '\n	');
  return contentWithSpaces ? `\n	${contentWithSpaces}` : '';
};

export interface MakefileProjenOptions {
  /**
   * Additional contents to be added to the Makefile on top of the default projen related targets
   * @default - no additional rules
   */
  readonly additionalMakefileContentsProjen?: string;
  /**
   * The version of ts-node lib to be used in "prepare" target of the Makefile
   * to install tooling for Projen to work
   * @default '10.9.2'
   */
  readonly tsNodeLibVersion?: string;
  /**
   * The version of projen lib to be used in "prepare" target of the Makefile
   * to install tooling for Projen to work
   * @default '0.91.6'
   */
  readonly projenLibVersion?: string;
  /**
   * Node version to be added to the .nvmrc file
   * @default '20.16.0'
   */
  readonly nodeVersion?: string;
  /**
   * Additional contents to be added to each target of the Makefile
   */
  readonly additionalMakefileContentsTargets?: Record<string, string>;
}
