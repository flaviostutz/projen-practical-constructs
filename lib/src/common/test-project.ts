import { mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
// eslint-disable-next-line unicorn/import-style
import { join } from 'node:path';

import { Project } from 'projen';

export class TestProject extends Project {
  constructor() {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    super({ name: 'test', outdir });
  }
}
