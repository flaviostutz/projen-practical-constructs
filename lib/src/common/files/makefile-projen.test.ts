/* eslint-disable no-new */
import { Testing } from 'projen';

import { TestProject } from '../test-project';

import { MakefileProjenFile } from './makefile-projen';

test('default Makefile for projen projects', () => {
  const project = new TestProject();
  new MakefileProjenFile(project);
  const output = Testing.synth(project);
  expect(output).toMatchSnapshot();
});
