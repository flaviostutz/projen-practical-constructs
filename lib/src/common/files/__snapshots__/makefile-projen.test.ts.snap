// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default Makefile for projen projects 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/Makefile-projen
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "Makefile-projen",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "Makefile-projen": "SHELL := /bin/bash

all: build lint test

prepare:
	brew install nvm
	@echo "Configure your shell following the instructions at https://formulae.brew.sh/formula/nvm"
  

prepare-projen:
	@if [ "$$CI" == "true" ]; then 		set -x; npm install --no-save --no-package-lock ts-node@10.9.2 projen@0.91.6; 	else 		set -x; npm install --no-save ts-node@10.9.2 projen@0.91.6; 	fi

# Remove projen from the project
eject:
	npx projen eject

# Prepare the project for compilation
pre-compile:
	npx projen pre-compile

# Only compile
compile:
	npx projen compile

# Runs after successful compilation
post-compile:
	npx projen post-compile

# Run tests
test:
	npx projen test

# Creates the distribution package
package:
	npx projen package

# Full release build
build:
	npx projen build

",
}
`;
